Buchanan-Allison-LiDAR-Uncertainty-Class (1).py

```diff
@@ -0,0 +1,274 @@
+        self.insitu_path = insitu_path
+        self.plots_path = plots_path
+        self.study_site = study_site
+
+    def calc_lidar_chm_stats(self):
+        '''
+        Calculates lidar mean and max tree heights.
+        '''
+
+        lidar_chm = rxr.open_rasterio(
+            self.lidar_path, masked=True).squeeze()
+        lidar_chm_clean = lidar_chm.where(lidar_chm > 0, np.nan)
+
+        plots_gdf = gpd.read_file(self.plots_path)
+        plots_gdf.geometry = plots_gdf.geometry.buffer(20)
+
+        lidar_chm_stats = rs.zonal_stats(
+            plots_gdf,
+            lidar_chm_clean.values,
+            affine=lidar_chm_clean.rio.transform(),
+            nodata=-999,
+            geojson_out=True,
+            copy_properties=True,
+            stats='count mean max')
+
+        # Create a gdf based on the stats requested from zonal stats and rename columns.
+        lidar_chm_gdf = gpd.GeoDataFrame.from_features(lidar_chm_stats)
+
+        lidar_chm_gdf.rename(
+            columns={'max': 'lidar_max',
+                     'mean': 'lidar_mean',
+                     'ID': 'plotid',
+                     'Plot_ID': 'plotid'}, inplace=True)
+
+        return lidar_chm_gdf
+
+    def calc_insitu_dfs(self):
+        '''
+        Calculate insitu tree height mean and max.
+        '''
+
+        df = (pd.read_csv(self.insitu_path)
+              .groupby('plotid')
+              .stemheight
+              .agg(['max', 'mean'])
+              .rename(columns={'max': 'insitu_max',
+                               'mean': 'insitu_mean',
+                               'ID': 'plotid',
+                               'Plot_ID': 'plotid'}))
+
+        return df
+
+    def merge_dfs(self):
+        '''
+        Merges insitu and lidar data frames together.
+
+        '''
+
+        chm_stats_gdf = self.calc_lidar_chm_stats()
+        insitu_height_df = self.calc_insitu_dfs()
+
+        if (self.study_site == 'soap'):
+            ss = self.study_site.upper()
+            chm_stats_gdf['plotid'] = ss + chm_stats_gdf['plotid']
+
+        data_frame = chm_stats_gdf.merge(
+            insitu_height_df, right_on='plotid', left_on='plotid')
+        return data_frame
+
+
+sjer_site = AnalyzeLidarUncertainty(
+    sjer_data_loader.chm_path, sjer_data_loader.insitu_path,
+    sjer_data_loader.plots_path, sjer_data_loader.formatting_dict['site_name_low'])
+
+sjer_site.calc_lidar_chm_stats()
+
+
+# In[84]:
+
+
+soap_site = AnalyzeLidarUncertainty(soap_data_loader.chm_path, soap_data_loader.insitu_path,
+                                    soap_data_loader.plots_path, soap_data_loader.formatting_dict['site_name_low'])
+soap_site.calc_lidar_chm_stats()
+
+
+# In[85]:
+
+
+sjer_site.calc_insitu_dfs()
+
+
+# In[86]:
+
+
+soap_site.calc_insitu_dfs()
+
+
+# In[87]:
+
+
+soap_site.merge_dfs()
+
+
+# In[88]:
+
+
+sjer_site.merge_dfs()
+
+
+# In[90]:
+
+
+def plot_regression(lidar_mean, insitu_mean, lidar_max, insitu_max):
```

Comment:
the code in the .py file for this section is different than the code in the .ipynb file... I'm not good enough yet at python to 'eyeball' which is better w/out running it. But... if you let me know which you intended to be the final version, happy to add more comments or help with any debugging. For either code, it looks like you could easily add site name as a parameter if you wanted to quickly get labels onto your plots. (Though there may be a more efficient method).